#Load Packages

In [1]:
import pandas as pd
import json
from datetime import datetime
import numpy as np
from ast import literal_eval

In [2]:
receipts = pd.read_json('receipts.json',lines=True)
users = pd.read_json('users.json',lines=True)
brands = pd.read_json('brands.json',lines=True)

# Parse Receipt Data

In [3]:
receipts.head()

_id  bonusPointsEarned  \
0  {'$oid': '5ff1e1eb0a720f0523000575'}              500.0   
1  {'$oid': '5ff1e1bb0a720f052300056b'}              150.0   
2  {'$oid': '5ff1e1f10a720f052300057a'}                5.0   
3  {'$oid': '5ff1e1ee0a7214ada100056f'}                5.0   
4  {'$oid': '5ff1e1d20a7214ada1000561'}                5.0   

                             bonusPointsEarnedReason  \
0  Receipt number 2 completed, bonus point schedu...   
1  Receipt number 5 completed, bonus point schedu...   
2                         All-receipts receipt bonus   
3                         All-receipts receipt bonus   
4                         All-receipts receipt bonus   

                 createDate               dateScanned  \
0  {'$date': 1609687531000}  {'$date': 1609687531000}   
1  {'$date': 1609687483000}  {'$date': 1609687483000}   
2  {'$date': 1609687537000}  {'$date': 1609687537000}   
3  {'$date': 1609687534000}  {'$date': 1609687534000}   
4  {'$date': 1609687506000}  {'$date': 1609687506000}   

               finishedDate                modifyDate  \
0  {'$date': 1609687531000}  {'$date': 1609687536000}   
1  {'$date': 1609687483000}  {'$date': 1609687488000}   
2                       NaN  {'$date': 1609687542000}   
3  {'$date': 1609687534000}  {'$date': 1609687539000}   
4  {'$date': 1609687511000}  {'$date': 1609687511000}   

          pointsAwardedDate  pointsEarned              purchaseDate  \
0  {'$date': 1609687531000}         500.0  {'$date': 1609632000000}   
1  {'$date': 1609687483000}         150.0  {'$date': 1609601083000}   
2                       NaN           5.0  {'$date': 1609632000000}   
3  {'$date': 1609687534000}           5.0  {'$date': 1609632000000}   
4  {'$date': 1609687506000}           5.0  {'$date': 1609601106000}   

   purchasedItemCount                             rewardsReceiptItemList  \
0                 5.0  [{'barcode': '4011', 'description': 'ITEM NOT ...   
1                 2.0  [{'barcode': '4011', 'description': 'ITEM NOT ...   
2                 1.0  [{'needsFetchReview': False, 'partnerItemId': ...   
3                 4.0  [{'barcode': '4011', 'description': 'ITEM NOT ...   
4                 2.0  [{'barcode': '4011', 'description': 'ITEM NOT ...   

  rewardsReceiptStatus  totalSpent                    userId  
0             FINISHED        26.0  5ff1e1eacfcf6c399c274ae6  
1             FINISHED        11.0  5ff1e194b6a9d73a3a9f1052  
2             REJECTED        10.0  5ff1e1f1cfcf6c399c274b0b  
3             FINISHED        28.0  5ff1e1eacfcf6c399c274ae6  
4             FINISHED         1.0  5ff1e194b6a9d73a3a9f1052

In [4]:
receipts['_id'] = receipts['_id'].apply(lambda x: x['$oid'])

In [5]:
def date_converter(x):
  try:
    return(datetime.utcfromtimestamp(int(x['$date'])/1000).strftime('%Y-%m-%d %H:%M:%S'))
  except TypeError:
    return(np.nan)

In [6]:
receipts['createDate'] = receipts['createDate'].apply(lambda x: date_converter(x))
receipts['dateScanned'] = receipts['dateScanned'].apply(lambda x: date_converter(x))
receipts['finishedDate'] = receipts['finishedDate'].apply(lambda x: date_converter(x))
receipts['modifyDate'] = receipts['modifyDate'].apply(lambda x: date_converter(x))
receipts['pointsAwardedDate'] = receipts['pointsAwardedDate'].apply(lambda x: date_converter(x))
receipts['purchaseDate'] = receipts['purchaseDate'].apply(lambda x: date_converter(x))

In [7]:
receipts.head()

_id  bonusPointsEarned  \
0  5ff1e1eb0a720f0523000575              500.0   
1  5ff1e1bb0a720f052300056b              150.0   
2  5ff1e1f10a720f052300057a                5.0   
3  5ff1e1ee0a7214ada100056f                5.0   
4  5ff1e1d20a7214ada1000561                5.0   

                             bonusPointsEarnedReason           createDate  \
0  Receipt number 2 completed, bonus point schedu...  2021-01-03 15:25:31   
1  Receipt number 5 completed, bonus point schedu...  2021-01-03 15:24:43   
2                         All-receipts receipt bonus  2021-01-03 15:25:37   
3                         All-receipts receipt bonus  2021-01-03 15:25:34   
4                         All-receipts receipt bonus  2021-01-03 15:25:06   

           dateScanned         finishedDate           modifyDate  \
0  2021-01-03 15:25:31  2021-01-03 15:25:31  2021-01-03 15:25:36   
1  2021-01-03 15:24:43  2021-01-03 15:24:43  2021-01-03 15:24:48   
2  2021-01-03 15:25:37                  NaN  2021-01-03 15:25:42   
3  2021-01-03 15:25:34  2021-01-03 15:25:34  2021-01-03 15:25:39   
4  2021-01-03 15:25:06  2021-01-03 15:25:11  2021-01-03 15:25:11   

     pointsAwardedDate  pointsEarned         purchaseDate  purchasedItemCount  \
0  2021-01-03 15:25:31         500.0  2021-01-03 00:00:00                 5.0   
1  2021-01-03 15:24:43         150.0  2021-01-02 15:24:43                 2.0   
2                  NaN           5.0  2021-01-03 00:00:00                 1.0   
3  2021-01-03 15:25:34           5.0  2021-01-03 00:00:00                 4.0   
4  2021-01-03 15:25:06           5.0  2021-01-02 15:25:06                 2.0   

                              rewardsReceiptItemList rewardsReceiptStatus  \
0  [{'barcode': '4011', 'description': 'ITEM NOT ...             FINISHED   
1  [{'barcode': '4011', 'description': 'ITEM NOT ...             FINISHED   
2  [{'needsFetchReview': False, 'partnerItemId': ...             REJECTED   
3  [{'barcode': '4011', 'description': 'ITEM NOT ...             FINISHED   
4  [{'barcode': '4011', 'description': 'ITEM NOT ...             FINISHED   

   totalSpent                    userId  
0        26.0  5ff1e1eacfcf6c399c274ae6  
1        11.0  5ff1e194b6a9d73a3a9f1052  
2        10.0  5ff1e1f1cfcf6c399c274b0b  
3        28.0  5ff1e1eacfcf6c399c274ae6  
4         1.0  5ff1e194b6a9d73a3a9f1052

In [8]:
rewardsReceiptItemList = receipts[['_id','rewardsReceiptItemList']]
receipts.drop(columns = 'rewardsReceiptItemList', inplace = True)

In [9]:
rewardsReceiptItemList = rewardsReceiptItemList.explode('rewardsReceiptItemList')
rewardsReceiptItemList.reset_index(inplace=True)

In [10]:
rewardsReceiptItemList.rewardsReceiptItemList = rewardsReceiptItemList.rewardsReceiptItemList.fillna('{}')

In [11]:
rewardsReceiptItemList.rewardsReceiptItemList = rewardsReceiptItemList.rewardsReceiptItemList.apply(lambda x:str(x))
rewardsReceiptItemList.rewardsReceiptItemList = rewardsReceiptItemList.rewardsReceiptItemList.apply(literal_eval)

In [12]:
rewardsReceiptItemList = rewardsReceiptItemList.join(pd.json_normalize(rewardsReceiptItemList.pop('rewardsReceiptItemList')))

In [13]:
rewardsReceiptItemList.head()

index                       _id       barcode  \
0      0  5ff1e1eb0a720f0523000575          4011   
1      1  5ff1e1bb0a720f052300056b          4011   
2      1  5ff1e1bb0a720f052300056b  028400642255   
3      2  5ff1e1f10a720f052300057a           NaN   
4      3  5ff1e1ee0a7214ada100056f          4011   

                                         description finalPrice itemPrice  \
0                                     ITEM NOT FOUND      26.00     26.00   
1                                     ITEM NOT FOUND          1         1   
2  DORITOS TORTILLA CHIP SPICY SWEET CHILI REDUCE...      10.00     10.00   
3                                                NaN        NaN       NaN   
4                                     ITEM NOT FOUND      28.00     28.00   

  needsFetchReview partnerItemId preventTargetGapPoints  quantityPurchased  \
0            False             1                   True                5.0   
1              NaN             1                    NaN                1.0   
2             True             2                   True                1.0   
3            False             1                   True                NaN   
4            False             1                   True                4.0   

   ... itemNumber originalMetaBriteQuantityPurchased pointsEarned  \
0  ...        NaN                                NaN          NaN   
1  ...        NaN                                NaN          NaN   
2  ...        NaN                                NaN          NaN   
3  ...        NaN                                NaN          NaN   
4  ...        NaN                                NaN          NaN   

   targetPrice competitiveProduct originalFinalPrice  \
0          NaN                NaN                NaN   
1          NaN                NaN                NaN   
2          NaN                NaN                NaN   
3          NaN                NaN                NaN   
4          NaN                NaN                NaN   

  originalMetaBriteItemPrice deleted priceAfterCoupon metabriteCampaignId  
0                        NaN     NaN              NaN                 NaN  
1                        NaN     NaN              NaN                 NaN  
2                        NaN     NaN              NaN                 NaN  
3                        NaN     NaN              NaN                 NaN  
4                        NaN     NaN              NaN                 NaN  

[5 rows x 36 columns]

In [14]:
receipts.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1119 entries, 0 to 1118
Data columns (total 14 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   _id                      1119 non-null   object 
 1   bonusPointsEarned        544 non-null    float64
 2   bonusPointsEarnedReason  544 non-null    object 
 3   createDate               1119 non-null   object 
 4   dateScanned              1119 non-null   object 
 5   finishedDate             568 non-null    object 
 6   modifyDate               1119 non-null   object 
 7   pointsAwardedDate        537 non-null    object 
 8   pointsEarned             609 non-null    float64
 9   purchaseDate             671 non-null    object 
 10  purchasedItemCount       635 non-null    float64
 11  rewardsReceiptStatus     1119 non-null   object 
 12  totalSpent               684 non-null    float64
 13  userId                   1119 non-null   object 
dtypes: float64(4), object(10

In [15]:
rewardsReceiptItemList.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7381 entries, 0 to 7380
Data columns (total 36 columns):
 #   Column                              Non-Null Count  Dtype  
---  ------                              --------------  -----  
 0   index                               7381 non-null   int64  
 1   _id                                 7381 non-null   object 
 2   barcode                             3090 non-null   object 
 3   description                         6560 non-null   object 
 4   finalPrice                          6767 non-null   object 
 5   itemPrice                           6767 non-null   object 
 6   needsFetchReview                    813 non-null    object 
 7   partnerItemId                       6941 non-null   object 
 8   preventTargetGapPoints              358 non-null    object 
 9   quantityPurchased                   6767 non-null   float64
 10  userFlaggedBarcode                  337 non-null    object 
 11  userFlaggedNewItem                  323 non

# Parse User Data

In [16]:
users.head()

_id  active               createdDate  \
0  {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
1  {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
2  {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   
3  {'$oid': '5ff1e1eacfcf6c399c274ae6'}    True  {'$date': 1609687530554}   
4  {'$oid': '5ff1e194b6a9d73a3a9f1052'}    True  {'$date': 1609687444800}   

                  lastLogin      role signUpSource state  
0  {'$date': 1609687537858}  consumer        Email    WI  
1  {'$date': 1609687537858}  consumer        Email    WI  
2  {'$date': 1609687537858}  consumer        Email    WI  
3  {'$date': 1609687530597}  consumer        Email    WI  
4  {'$date': 1609687537858}  consumer        Email    WI

In [17]:
users['_id'] = users['_id'].apply(lambda x: x['$oid'])
users['createdDate'] = users['createdDate'].apply(lambda x: date_converter(x))
users['lastLogin'] = users['lastLogin'].apply(lambda x: date_converter(x))

In [18]:
users.head()

_id  active          createdDate            lastLogin  \
0  5ff1e194b6a9d73a3a9f1052    True  2021-01-03 15:24:04  2021-01-03 15:25:37   
1  5ff1e194b6a9d73a3a9f1052    True  2021-01-03 15:24:04  2021-01-03 15:25:37   
2  5ff1e194b6a9d73a3a9f1052    True  2021-01-03 15:24:04  2021-01-03 15:25:37   
3  5ff1e1eacfcf6c399c274ae6    True  2021-01-03 15:25:30  2021-01-03 15:25:30   
4  5ff1e194b6a9d73a3a9f1052    True  2021-01-03 15:24:04  2021-01-03 15:25:37   

       role signUpSource state  
0  consumer        Email    WI  
1  consumer        Email    WI  
2  consumer        Email    WI  
3  consumer        Email    WI  
4  consumer        Email    WI

In [19]:
users.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 495 entries, 0 to 494
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   _id           495 non-null    object
 1   active        495 non-null    bool  
 2   createdDate   495 non-null    object
 3   lastLogin     433 non-null    object
 4   role          495 non-null    object
 5   signUpSource  447 non-null    object
 6   state         439 non-null    object
dtypes: bool(1), object(6)
memory usage: 23.8+ KB


# Parse Brand Data

In [20]:
brands.head()

_id       barcode        category  \
0  {'$oid': '601ac115be37ce2ead437551'}  511111019862          Baking   
1  {'$oid': '601c5460be37ce2ead43755f'}  511111519928       Beverages   
2  {'$oid': '601ac142be37ce2ead43755d'}  511111819905          Baking   
3  {'$oid': '601ac142be37ce2ead43755a'}  511111519874          Baking   
4  {'$oid': '601ac142be37ce2ead43755e'}  511111319917  Candy & Sweets   

       categoryCode                                                cpg  \
0            BAKING  {'$id': {'$oid': '601ac114be37ce2ead437550'}, ...   
1         BEVERAGES  {'$id': {'$oid': '5332f5fbe4b03c9a25efd0ba'}, ...   
2            BAKING  {'$id': {'$oid': '601ac142be37ce2ead437559'}, ...   
3            BAKING  {'$id': {'$oid': '601ac142be37ce2ead437559'}, ...   
4  CANDY_AND_SWEETS  {'$id': {'$oid': '5332fa12e4b03c9a25efd1e7'}, ...   

                        name  topBrand                      brandCode  
0  test brand @1612366101024       0.0                            NaN  
1                  Starbucks       0.0                      STARBUCKS  
2  test brand @1612366146176       0.0  TEST BRANDCODE @1612366146176  
3  test brand @1612366146051       0.0  TEST BRANDCODE @1612366146051  
4  test brand @1612366146827       0.0  TEST BRANDCODE @1612366146827

In [21]:
brands['_id'] = brands['_id'].apply(lambda x: x['$oid'])

In [22]:
cpg = brands[['_id','cpg']]
brands.drop(columns = 'cpg', inplace = True)

In [23]:
cpg = cpg.join(pd.json_normalize(cpg.pop('cpg')))

In [24]:
cpg

_id  $ref                  $id.$oid
0     601ac115be37ce2ead437551  Cogs  601ac114be37ce2ead437550
1     601c5460be37ce2ead43755f  Cogs  5332f5fbe4b03c9a25efd0ba
2     601ac142be37ce2ead43755d  Cogs  601ac142be37ce2ead437559
3     601ac142be37ce2ead43755a  Cogs  601ac142be37ce2ead437559
4     601ac142be37ce2ead43755e  Cogs  5332fa12e4b03c9a25efd1e7
...                        ...   ...                       ...
1162  5f77274dbe37ce6b592e90c0  Cogs  5f77274dbe37ce6b592e90bf
1163  5dc1fca91dda2c0ad7da64ae  Cogs  53e10d6368abd3c7065097cc
1164  5f494c6e04db711dd8fe87e7  Cogs  5332fa12e4b03c9a25efd1e7
1165  5a021611e4b00efe02b02a57  Cogs  5332f5f6e4b03c9a25efd0b4
1166  6026d757be37ce6369301468  Cogs  6026d757be37ce6369301467

[1167 rows x 3 columns]

In [25]:
cpg.rename(columns={"$ref": "ref", "$id.$oid": "oid"}, inplace = True)

In [26]:
cpg

_id   ref                       oid
0     601ac115be37ce2ead437551  Cogs  601ac114be37ce2ead437550
1     601c5460be37ce2ead43755f  Cogs  5332f5fbe4b03c9a25efd0ba
2     601ac142be37ce2ead43755d  Cogs  601ac142be37ce2ead437559
3     601ac142be37ce2ead43755a  Cogs  601ac142be37ce2ead437559
4     601ac142be37ce2ead43755e  Cogs  5332fa12e4b03c9a25efd1e7
...                        ...   ...                       ...
1162  5f77274dbe37ce6b592e90c0  Cogs  5f77274dbe37ce6b592e90bf
1163  5dc1fca91dda2c0ad7da64ae  Cogs  53e10d6368abd3c7065097cc
1164  5f494c6e04db711dd8fe87e7  Cogs  5332fa12e4b03c9a25efd1e7
1165  5a021611e4b00efe02b02a57  Cogs  5332f5f6e4b03c9a25efd0b4
1166  6026d757be37ce6369301468  Cogs  6026d757be37ce6369301467

[1167 rows x 3 columns]

In [27]:
brands.topBrand.unique()

array([ 0., nan,  1.])

In [28]:
brands.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1167 entries, 0 to 1166
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   _id           1167 non-null   object 
 1   barcode       1167 non-null   int64  
 2   category      1012 non-null   object 
 3   categoryCode  517 non-null    object 
 4   name          1167 non-null   object 
 5   topBrand      555 non-null    float64
 6   brandCode     933 non-null    object 
dtypes: float64(1), int64(1), object(5)
memory usage: 63.9+ KB


In [29]:
cpg.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1167 entries, 0 to 1166
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   _id     1167 non-null   object
 1   ref     1167 non-null   object
 2   oid     1167 non-null   object
dtypes: object(3)
memory usage: 27.5+ KB


# SQL

In [30]:
pip install pandasql

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [31]:
import pandasql as ps

In [32]:
receipts.rewardsReceiptStatus.unique()

array(['FINISHED', 'REJECTED', 'FLAGGED', 'SUBMITTED', 'PENDING'],
      dtype=object)

In [33]:
receipts.loc[receipts['rewardsReceiptStatus'] == 'FINISHED'].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 518 entries, 0 to 1021
Data columns (total 14 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   _id                      518 non-null    object 
 1   bonusPointsEarned        456 non-null    float64
 2   bonusPointsEarnedReason  456 non-null    object 
 3   createDate               518 non-null    object 
 4   dateScanned              518 non-null    object 
 5   finishedDate             518 non-null    object 
 6   modifyDate               518 non-null    object 
 7   pointsAwardedDate        514 non-null    object 
 8   pointsEarned             518 non-null    float64
 9   purchaseDate             518 non-null    object 
 10  purchasedItemCount       518 non-null    float64
 11  rewardsReceiptStatus     518 non-null    object 
 12  totalSpent               518 non-null    float64
 13  userId                   518 non-null    object 
dtypes: float64(4), object(10)

In [34]:
receipts.loc[receipts['rewardsReceiptStatus'] == 'REJECTED'].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 71 entries, 2 to 1117
Data columns (total 14 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   _id                      71 non-null     object 
 1   bonusPointsEarned        58 non-null     float64
 2   bonusPointsEarnedReason  58 non-null     object 
 3   createDate               71 non-null     object 
 4   dateScanned              71 non-null     object 
 5   finishedDate             0 non-null      object 
 6   modifyDate               71 non-null     object 
 7   pointsAwardedDate        4 non-null      object 
 8   pointsEarned             58 non-null     float64
 9   purchaseDate             69 non-null     object 
 10  purchasedItemCount       71 non-null     float64
 11  rewardsReceiptStatus     71 non-null     object 
 12  totalSpent               71 non-null     float64
 13  userId                   71 non-null     object 
dtypes: float64(4), object(10)


In [35]:
receipts.loc[receipts['rewardsReceiptStatus'] == 'FLAGGED'].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 46 entries, 12 to 755
Data columns (total 14 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   _id                      46 non-null     object 
 1   bonusPointsEarned        30 non-null     float64
 2   bonusPointsEarnedReason  30 non-null     object 
 3   createDate               46 non-null     object 
 4   dateScanned              46 non-null     object 
 5   finishedDate             0 non-null      object 
 6   modifyDate               46 non-null     object 
 7   pointsAwardedDate        19 non-null     object 
 8   pointsEarned             33 non-null     float64
 9   purchaseDate             35 non-null     object 
 10  purchasedItemCount       46 non-null     float64
 11  rewardsReceiptStatus     46 non-null     object 
 12  totalSpent               46 non-null     float64
 13  userId                   46 non-null     object 
dtypes: float64(4), object(10)


In [36]:
receipts.loc[receipts['rewardsReceiptStatus'] == 'SUBMITTED'].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 434 entries, 71 to 1118
Data columns (total 14 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   _id                      434 non-null    object 
 1   bonusPointsEarned        0 non-null      float64
 2   bonusPointsEarnedReason  0 non-null      object 
 3   createDate               434 non-null    object 
 4   dateScanned              434 non-null    object 
 5   finishedDate             0 non-null      object 
 6   modifyDate               434 non-null    object 
 7   pointsAwardedDate        0 non-null      object 
 8   pointsEarned             0 non-null      float64
 9   purchaseDate             0 non-null      object 
 10  purchasedItemCount       0 non-null      float64
 11  rewardsReceiptStatus     434 non-null    object 
 12  totalSpent               0 non-null      float64
 13  userId                   434 non-null    object 
dtypes: float64(4), object(10

In [37]:
receipts.loc[receipts['rewardsReceiptStatus'] == 'PENDING'].info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 50 entries, 210 to 875
Data columns (total 14 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   _id                      50 non-null     object 
 1   bonusPointsEarned        0 non-null      float64
 2   bonusPointsEarnedReason  0 non-null      object 
 3   createDate               50 non-null     object 
 4   dateScanned              50 non-null     object 
 5   finishedDate             50 non-null     object 
 6   modifyDate               50 non-null     object 
 7   pointsAwardedDate        0 non-null      object 
 8   pointsEarned             0 non-null      float64
 9   purchaseDate             49 non-null     object 
 10  purchasedItemCount       0 non-null      float64
 11  rewardsReceiptStatus     50 non-null     object 
 12  totalSpent               49 non-null     float64
 13  userId                   50 non-null     object 
dtypes: float64(4), object(10)

In [38]:
receipts.loc[receipts['rewardsReceiptStatus'] == 'PENDING']

_id  bonusPointsEarned bonusPointsEarnedReason  \
210  5ffe1cbe0a7214ad28002843                NaN                     NaN   
211  5ffce7db0a720f051500236e                NaN                     NaN   
213  5ffce8310a7214ad4e003797                NaN                     NaN   
214  5ffe19d90a7214ad28000e62                NaN                     NaN   
215  5ffe1d030a720f05ac002c9e                NaN                     NaN   
218  5ffce76e0a720f0515000b48                NaN                     NaN   
219  5ffe23560a720f05ac006874                NaN                     NaN   
220  5ffe22a20a720f05ac0061d7                NaN                     NaN   
221  5ffe1dc20a7214ad28003180                NaN                     NaN   
224  5ffe1aa90a7214ad280015e8                NaN                     NaN   
225  5ffcef5a0a720f0515004f1a                NaN                     NaN   
226  5ffe23d70a7214ad280068f6                NaN                     NaN   
227  5ffe18b10a7214ad280003d1                NaN                     NaN   
228  5ffcef630a720f05150050ae                NaN                     NaN   
231  5ffe20290a720f05ac004a84                NaN                     NaN   
232  5ffe1c520a7214ad2800245c                NaN                     NaN   
233  5ffe1b630a7214ad28001c79                NaN                     NaN   
235  5ffe1fa40a7214ad2800427f                NaN                     NaN   
237  5ffe23cf0a720f05ac006cc2                NaN                     NaN   
238  5ffcef570a720f0515004e91                NaN                     NaN   
555  601431e40a7214ad50000041                NaN                     NaN   
759  601d7c000a7214ad590002a1                NaN                     NaN   
760  601d7c180a720f05540002fd                NaN                     NaN   
761  601d77170a720f053c00000a                NaN                     NaN   
762  601d7b8e0a720f0554000079                NaN                     NaN   
763  601d7d580a7214ad59000928                NaN                     NaN   
764  601d7cf90a720f055400070e                NaN                     NaN   
765  601d77180a720f053c00000f                NaN                     NaN   
766  601d7d540a720f05540008a6                NaN                     NaN   
767  601d7ba00a720f05540000d2                NaN                     NaN   
768  601d7bb60a7214ad59000141                NaN                     NaN   
769  601d7cd30a720f0554000663                NaN                     NaN   
770  601d7b730a720f0554000006                NaN                     NaN   
771  601d7d340a7214ad5900086c                NaN                     NaN   
772  601d7c1e0a720f055400031a                NaN                     NaN   
774  601d7c300a7214ad59000384                NaN                     NaN   
775  601d7ca40a720f055400058b                NaN                     NaN   
776  601d7b770a720f0554000016                NaN                     NaN   
777  601d7bb60a720f055400013a                NaN                     NaN   
778  601d7c7f0a7214ad590004f2                NaN                     NaN   
779  601d7d2c0a720f05540007fa                NaN                     NaN   
780  601d7d230a720f05540007ce                NaN                     NaN   
781  601d7c500a720f0554000409                NaN                     NaN   
783  601d7c960a7214ad59000564                NaN                     NaN   
784  601d75dd0a720f05a9000004                NaN                     NaN   
785  601d7c480a720f05540003e0                NaN                     NaN   
786  601d7c280a7214ad5900035c                NaN                     NaN   
787  601d7cd50a7214ad5900069f                NaN                     NaN   
788  601d7b840a720f055400004e                NaN                     NaN   
875  602160a30a720f05c0000001                NaN                     NaN   

              createDate          dateScanned         finishedDate  \
210  2021-01-12 22:03:42  2021-01-12 22:03:42  2021-01-12 22:03:42   
211  2021

In [39]:
result = ps.sqldf(
    " select average_spend_Accepted, average_spend_Rejected, total_number_items_Accepted, total_number_items_Rejected from (select sum(totalSpent)/count(_id) as average_spend_Accepted, sum(purchasedItemCount) as total_number_items_Accepted from receipts where rewardsReceiptStatus == 'FINISHED' or rewardsReceiptStatus == 'FLAGGED') as a cross join (select sum(totalSpent)/count(_id) as average_spend_Rejected, sum(purchasedItemCount) as total_number_items_Rejected from receipts where rewardsReceiptStatus == 'REJECTED') as b"
)


In [40]:
result

average_spend_Accepted  average_spend_Rejected  \
0                 88.9775               23.326056   

   total_number_items_Accepted  total_number_items_Rejected  
0                       9198.0                        173.0

# Data Quality Issues Evaluation

Investigation of null values in lastLogin from users table

In [49]:
users.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 495 entries, 0 to 494
Data columns (total 7 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   _id           495 non-null    object
 1   active        495 non-null    bool  
 2   createdDate   495 non-null    object
 3   lastLogin     433 non-null    object
 4   role          495 non-null    object
 5   signUpSource  447 non-null    object
 6   state         439 non-null    object
dtypes: bool(1), object(6)
memory usage: 23.8+ KB


In [52]:
consumer = users.loc[users['role'] == 'consumer']

In [64]:
consumer.nunique()

_id             204
active            2
createdDate     204
lastLogin       163
role              1
signUpSource      2
state             7
dtype: int64

Among 204 unique consumer ids, there are 41 entries do not have the last login record. These NaNs indicate that the consumer might become inactive. 

After going through Fetch Reward phone application's user journey, I think failing the receipt validation process might be a user pain point that causes users to churn and leave null values in users table.

Step1 1: Store NaN lastLogin userIDs from users table into a list

In [55]:
users_churn = consumer.loc[consumer['lastLogin'].isna()]

In [56]:
users_churn

_id  active          createdDate lastLogin  \
97   5ff616a68f142f11dd189163    True  2021-01-06 19:59:34       NaN   
143  5ffe115404929101d0aaebb2    True  2021-01-12 21:15:00       NaN   
148  5ffe115404929101d0aaebb2    True  2021-01-12 21:15:00       NaN   
170  5e27526d0bdb6a138c32b556    True  2020-01-21 19:35:09       NaN   
180  6002475cfb296c121a81b98d    True  2021-01-16 01:54:36       NaN   
..                        ...     ...                  ...       ...   
381  60186237c8b50e11d8454d5f    True  2021-02-01 20:19:03       NaN   
382  60186237c8b50e11d8454d5f    True  2021-02-01 20:19:03       NaN   
389  60217fa799409b11fcf899fe    True  2021-02-08 18:15:03       NaN   
420  5fb0a078be5fc9775c1f3945    True  2020-11-15 03:28:56       NaN   
429  5fb0a078be5fc9775c1f3945    True  2020-11-15 03:28:56       NaN   

         role signUpSource state  
97   consumer        Email    KY  
143  consumer        Email    AL  
148  consumer        Email    AL  
170  consumer       Google    WI  
180  consumer        Email    WI  
..        ...          ...   ...  
381  consumer        Email   NaN  
382  consumer        Email   NaN  
389  consumer        Email    WI  
420  consumer       Google    AL  
429  consumer       Google    AL  

[62 rows x 7 columns]

In [ ]:
with pd.option_context("display.max_rows", 1000):
    display(users_churn)

In [58]:
churn_lst = users_churn['_id'].tolist()

Step 2: Use [churn_lst] to locate [_id] in receipts table

In [59]:
churn_receipt = receipts.loc[receipts['userId'].isin(churn_lst)]

Step 3: Filter out accepted rewards

In [60]:
churn_receipt_rejected = churn_receipt.loc[(churn_receipt['rewardsReceiptStatus'] != 'FINISHED') & (churn_receipt['rewardsReceiptStatus'] != 'FLAGGED')]

In [61]:
churn_receipt_rejected

_id  bonusPointsEarned  \
246  5ffe17e20a720f05ac0000e7              750.0   
300  600205f60a7214ad4c00007b                NaN   
302  6001bb110a720f05f3000085                NaN   
307  600247cd0a7214ad4c00007e                NaN   
308  60025eb70a720f05f300008e                NaN   
311  600254550a720f05f300008c                NaN   
312  60024fa00a7214ad4c00007f                NaN   
313  60025f550a7214ad4c000084                NaN   
319  600498e60a720f05f3000093                NaN   
322  600496350a7214ad4c000087                NaN   
418  6008ea160a720f05fa000126                NaN   
432  600b89630a720f0535000011                NaN   
645  601868260a720f05f400000e                NaN   
661  6018682c0a720f05f4000011                NaN   
671  6018b92a0a7214ad2800005e                NaN   
674  6018b93c0a7214ad28000061                NaN   

                               bonusPointsEarnedReason           createDate  \
246  Receipt number 1 completed, bonus point schedu...  2021-01-12 21:42:58   
300                                                NaN  2021-01-15 21:15:32   
302                                                NaN  2021-01-15 15:56:01   
307                                                NaN  2021-01-16 01:56:29   
308                                                NaN  2021-01-16 03:34:15   
311                                                NaN  2021-01-16 02:49:57   
312                                                NaN  2021-01-16 02:29:52   
313                                                NaN  2021-01-16 03:36:53   
319                                                NaN  2021-01-17 20:07:02   
322                                                NaN  2021-01-17 19:55:33   
418                                                NaN  2021-01-21 02:42:30   
432                                                NaN  2021-01-23 02:26:43   
645                                                NaN  2021-02-01 20:44:22   
661                                                NaN  2021-02-01 20:44:28   
671                                                NaN  2021-02-02 02:30:02   
674                                                NaN  2021-02-02 02:30:20   

             dateScanned finishedDate           modifyDate pointsAwardedDate  \
246  2021-01-12 21:42:58          NaN  2021-01-13 16:00:54               NaN   
300  2021-01-15 21:15:32          NaN  2021-01-15 21:15:32               NaN   
302  2021-01-15 15:56:01          NaN  2021-01-15 15:56:01               NaN   
307  2021-01-16 01:56:29          NaN  2021-01-16 01:56:29               NaN   
308  2021-01-16 03:34:15          NaN  2021-01-16 03:34:15               NaN   
311  2021-01-16 02:49:57          NaN  2021-01-16 02:49:57               NaN   
312  2021-01-16 02:29:52          NaN  2021-01-16 02:29:52               NaN   
313  2021-01-16 03:36:53          NaN  2021-01-16 03:36:53               NaN   
319  2021-01-17 20:07:02          NaN  2021-01-17 20:07:02               NaN   
322  2021-01-17 19:55:33          NaN  2021-01-17 19:55:33               NaN   
418  2021-01-21 02:42:30          NaN  2021-01-21 02:42:30               NaN   
432  2021-01-23 02:26:43          NaN  2021-01-23 02:26:43               NaN   
645  2021-02-01 20:44:22          NaN  2021-02-01 20:44:22               NaN   
661  2021-02-01 20:44:28          NaN  2021-02-01 20:44:28               NaN   
671  2021-02-02 02:30:02          NaN  2021-02-02 02:30:02               NaN   
674  2021-02-02 02:30:20          NaN  2021-02-02 02:30:20               NaN   

     pointsEarned         purchaseDate  purchasedItemCount  \
246         750.0  2021-01-12 00:00:00                 1.0   
300           NaN                  NaN                 NaN   
302           NaN                  NaN                 NaN   
307           NaN                  NaN                 NaN   
308           NaN                  NaN                 NaN   
311           NaN                  NaN                 NaN   
31

In [62]:
churn_rejected_id = churn_receipt_rejected['_id'].tolist()

In [63]:
rewardsReceiptItemList.loc[rewardsReceiptItemList['_id'].isin(churn_rejected_id)] 

index                       _id       barcode  \
569     246  5ffe17e20a720f05ac0000e7  072785048119   
716     300  600205f60a7214ad4c00007b           NaN   
800     302  6001bb110a720f05f3000085           NaN   
1179    307  600247cd0a7214ad4c00007e           NaN   
1180    308  60025eb70a720f05f300008e           NaN   
1402    311  600254550a720f05f300008c           NaN   
1403    312  60024fa00a7214ad4c00007f           NaN   
1404    313  60025f550a7214ad4c000084           NaN   
1927    319  600498e60a720f05f3000093           NaN   
2114    322  600496350a7214ad4c000087           NaN   
3063    418  6008ea160a720f05fa000126           NaN   
4588    432  600b89630a720f0535000011           NaN   
6804    645  601868260a720f05f400000e           NaN   
6822    661  6018682c0a720f05f4000011           NaN   
6842    671  6018b92a0a7214ad2800005e           NaN   
6845    674  6018b93c0a7214ad28000061           NaN   

                           description finalPrice itemPrice needsFetchReview  \
569   Germ-X Hand Sanitizer - Original       2.29      2.29              NaN   
716                                NaN        NaN       NaN              NaN   
800                                NaN        NaN       NaN              NaN   
1179                               NaN        NaN       NaN              NaN   
1180                               NaN        NaN       NaN              NaN   
1402                               NaN        NaN       NaN              NaN   
1403                               NaN        NaN       NaN              NaN   
1404                               NaN        NaN       NaN              NaN   
1927                               NaN        NaN       NaN              NaN   
2114                               NaN        NaN       NaN              NaN   
3063                               NaN        NaN       NaN              NaN   
4588                               NaN        NaN       NaN              NaN   
6804                               NaN        NaN       NaN              NaN   
6822                               NaN        NaN       NaN              NaN   
6842                               NaN        NaN       NaN              NaN   
6845                               NaN        NaN       NaN              NaN   

     partnerItemId preventTargetGapPoints  quantityPurchased  ... itemNumber  \
569              0                    NaN                1.0  ...        NaN   
716            NaN                    NaN                NaN  ...        NaN   
800            NaN                    NaN                NaN  ...        NaN   
1179           NaN                    NaN                NaN  ...        NaN   
1180           NaN                    NaN                NaN  ...        NaN   
1402           NaN                    NaN                NaN  ...        NaN   
1403           NaN                    NaN                NaN  ...        NaN   
1404           NaN                    NaN                NaN  ...        NaN   
1927           NaN                    NaN                NaN  ...        NaN   
2114           NaN                    NaN                NaN  ...        NaN   
3063           NaN                    NaN                NaN  ...        NaN   
4588           NaN                    NaN                NaN  ...        NaN   
6804           NaN                    NaN                NaN  ...        NaN   
6822           NaN                    NaN                NaN  ...        NaN   
6842           NaN                    NaN                NaN  ...        NaN   
6845           NaN                    NaN                NaN  ...        NaN   

     originalMetaBriteQuantityPurchased pointsEarned  targetPrice  \
569                                 NaN          NaN          NaN   
716                                 NaN          NaN          NaN   
800                                 NaN          NaN          NaN   
1179                                NaN          NaN          NaN   
118

Finding: This thread leads to a matrix of NaNs (with one outlier) in the rewardsReceiptItemList table. More evidence is needed to support that reward validation rejection is an actual cause of null values in users table. 